### PR TITLE
Add extension field to image class

### PR DESCRIPTION
**Pull request by ahojnnes**

### CODE REVIEW

**Comment by B1ueber2y** [comment]
Thanks! This is great and very helpful. One potential risk of the design is the deprecation of operator == and != when the extension is available in both image objects.

**Comment by ahojnnes** [comment]
I would not compare the extensions. This is left to the user.

File: src/colmap/scene/image.h [hunk 1]
```diff
@@ -242,11 +259,27 @@ const std::vector<struct Point2D>& Image::Points2D() const { return points2D_; }
...
+bool Image::HasExtension() const { return extension_ != nullptr; }
+
+template <typename T>
+T& Image::GetExtension() {
+  return *static_cast<T*>(extension_.get());
+}
+
+template <typename T>
+const T& Image::GetExtension() const {
+  return *static_cast<const T*>(extension_.get());
```
[the remark this answers is not in the document]
Comment (B1ueber2y) [reply src/colmap/scene/image.h:280]: Addressed in https://github.com/colmap/colmap/pull/3015/commits/7a64347d218c9d4aea12c83ce8a0253aff834123. Will add a python example test in a separate test branch on this (not to be merged).

[the remark this answers is not in the document]
Comment (B1ueber2y) [reply src/colmap/scene/image.h:280]: Added a test pybind in a test branch: https://github.com/B1ueber2y/colmap/commit/a6cca3c1a753dc0d7e7e860a4332d1cdb4f5b08b (not to be merged).

Example with the pycolmap built from the test branch:
```
import pycolmap
image = pycolmap.create_test_image()
print(type(image.extension))
print(image.extension.test_field)
```
Output:
```
<class `pycolmap._core.TestExtension`>
42
```

This verifies that using the direct getter for std::shared_ptr<Extension> will enable automatic dynamic casting with pybind11.

**Comment by B1ueber2y** [comment]
Happy with the newest update from my side. Shall we move forward? : )

**Review by B1ueber2y — APPROVED** [review]
The current state works perfectly for me. I could not request reviews from your side as this is a PR created by you, but would be great if you could have a double check : )

**Review by sarlinpe — APPROVED** [review]

**Pull request closed without merging**